### PR TITLE
feature: implemented the ngx.re.gmatch function with FFI.

### PR DESCRIPTION
**Pull request by spacewander**
In some naive benchmark cases, the FFI(jittable) version is over 30% faster than the original implementation(in lua-nginx-module).

Looking forward for any further discussion about the new implementation.

I hereby granted the copyright of the changes in this pull request
to the authors of this lua-nginx-module project.

### CODE REVIEW

**Comment by agentzh** [comment]
@spacewander Oh, this is nice!

@doujiang24 @dndx @thibaultcha Will you review this PR for me? Many thanks!

**Comment by spacewander** [comment]
Any other response here?

**Comment by spacewander** [comment]
Just provide more info for those who are curious about **over 30% faster than the original implementation**:

#### Case 1

This case is taken directly from the docs:
```lua
local old = ngx.now()
for _ = 1, 1e6 do
    local iterator, err = ngx.re.gmatch("hello, world!", "([a-z]+)", "i")
    if not iterator then
        ngx.log(ngx.ERR, "error: ", err)
        return
    end

    local m
    m, err = iterator()    -- m[0] == m[1] == "hello"
    if err then
        ngx.log(ngx.ERR, "error: ", err)
        return
    end

    m, err = iterator()    -- m[0] == m[1] == "world"
    if err then
        ngx.log(ngx.ERR, "error: ", err)
        return
    end

    m, err = iterator()    -- m == nil
    if err then
        ngx.log(ngx.ERR, "error: ", err)
        return
    end
end
ngx.update_time()
ngx.say(ngx.now() - old)
```

The average time cost is:

* Original: 3.7069997787476
* New: 2.8710000514984

Only 29% faster. No so faster, isn't it?

#### Case 2

This case is taken from our code running in the production(with some modification, so it looks silly):

```lua
local old = ngx.now()

local data = table.concat({
    'md5s=md5s_value',
    'ext=ext_value',
    'mid=mid_value',
}, '\r\n')
local regex = [[(\w+)=([\S\s]*?)\r\n]]
for _ = 1, 1e6 do
    for m in ngx.re.gmatch(data, regex, "jo") do
        local name = m[1]
        local value = m[2]
     end
end
ngx.update_time()
ngx.say(ngx.now() - old)
```

The average time cost is:

* Original: 1.7410001754761
* New: 1.1369998455048

53% faster! That's fantastic.
Thanks for the power of JIT!

#### Conclusion

IMHO, most cases would be laid between case 1 and case 2.
 So **over 30% faster than the original implementation** is a reasonable announcement.

**Comment by spacewander** [comment]
@agentzh @doujiang24 @dndx @thibaultcha
Do you have time to take a look at this pr? It could be a gift for those people use `ngx.re.gmatch` in their project. Any feedback is welcome.

**Comment by agentzh** [comment]
@spacewander Yeah, we'll look into this. Sorry for the delay!

File: lib/resty/core/regex.lua [hunk 1]
```diff
@@ -479,6 +480,71 @@ function ngx.re.find(subj, regex, opts, ctx, nth)
...
+    local compiled = self._compiled
+    local subj = self._subj
+    local flags = self._flags
+    local pos = self._pos
+    -- The iterator is exhausted.
+    if not pos then
+        return nil
+    end
+
+    local rc = C.ngx_http_lua_ffi_exec_regex(compiled, flags, subj, #subj, pos)
```
Comment (thibaultcha) [inline lib/resty/core/regex.lua]: Maybe we could cache the subject length in `self._subj_len` to avoid calling the length operator on each iteration?

File: lib/resty/core/regex.lua [hunk 2]
```diff
@@ -479,6 +480,71 @@ function ngx.re.find(subj, regex, opts, ctx, nth)
...
+        if band(flags, FLAG_DFA) == 0 then
+            self._pos = nil
+            return nil, "capture size too small"
+        end
+
+        rc = 1
+    end
+
+    self._pos = compiled.captures[1]
+    local res = collect_captures(compiled, rc, subj, flags)
```
Comment (thibaultcha) [inline lib/resty/core/regex.lua]: I was under the impression that this was done to avoid `return collect_captures()` from returning multiple values, but it seems `collect_captures` only returns one value anyways. Maybe we could spare ourselves the variable here and `return collect_capture()` as a tail call? Or any reason why not?

File: t/re-gmatch.t [hunk 3]
```diff
@@ -0,0 +1,451 @@
...
+        }
+    }
+--- request
+GET /re
+--- response_body
+matched: hello
+matched: nil
+matched: nil
+--- no_error_log
+[error]
```
Comment (thibaultcha) [inline t/re-gmatch.t]: I was under the impression that this library's tests were supposed to test the JIT capabilities of the Lua code only, although, I do see that most of those tests come from `re-match.t`. No opinion on the matter, just saying.

However, what those tests are missing I believe, are test cases related to potential errors, as can be found in `035-gmatch.t` of the lua-nginx-module test suite. Errors from `re_match_compile` and `lua_ffi_exec_regex`. Should we add those, or will they be run from the lua-nginx-module test suite with this implementation? Quid if one does not run those tests from lua-nginx-module?

File: lib/resty/core/regex.lua [hunk 4]
```diff
@@ -479,6 +480,71 @@ function ngx.re.find(subj, regex, opts, ctx, nth)
 end
 
 
+local function iterate_re_gmatch(self)
+    local compile_once = self._compile_once
```
Comment (thibaultcha) [inline lib/resty/core/regex.lua]: Is it worth it to dereference this variable on each iteration, considering we will only use it, once, on the last iteration?

File: t/re-gmatch.t (hunk 3 above)
Comment (spacewander) [inline t/re-gmatch.t]: @thibaultcha 
IMHO, errors from `re_match_compile` and `ngx_http_lua_ffi_exec_regex` do not strictly belong to `ngx.re.gmatch`. Maybe we could add a `re-base.t` later and put the test cases there?

Comment (thibaultcha) [inline t/re-gmatch.t]: Yes, but one must also make sure they are properly handled and returned by `ngx.re.gmatch`, is what I meant.

**Comment by spacewander** [comment]
@thibaultcha @agentzh 
Thanks for the code review! Code updated.

**Comment by spacewander** [comment]
Just add `re-base.t` to ensure potential regex errors are handled well, and change `.travis.yml` to build PCRE libary with UTF8 support.

File: t/re-gmatch.t (hunk 3 above)
Comment (spacewander) [inline t/re-gmatch.t]: OK, I have added `re-base.t` to take care of them.

**Comment by agentzh** [comment]
@spacewander This is still in my review queue. I'll look at this carefully as soon as I can manage. Thanks for your patience!

File: lib/resty/core/regex.lua [hunk 5]
```diff
@@ -479,6 +480,72 @@ function ngx.re.find(subj, regex, opts, ctx, nth)
...
+
+    local re_gmatch_iterator = new_tab(0, 6)
+    re_gmatch_iterator._compiled = compiled
+    re_gmatch_iterator._compile_once = compile_once
+    re_gmatch_iterator._subj = subj
+    re_gmatch_iterator._subj_len = #subj
+    re_gmatch_iterator._flags = flags
+    re_gmatch_iterator._pos = 0
+
+    return setmetatable(re_gmatch_iterator, { __call = iterate_re_gmatch })
```
Comment (thibaultcha) [inline lib/resty/core/regex.lua]: It would be a nice improvement too to avoid recreating a new metatable for each call to `re.gmatch`. It can live as an upvalue and avoid unnecessary pressure on the GC:
```lua
do
    local re_gmatch_iterator_mt = { __call = iterate_re_gmatch }

    function ngx.re.gmatch(subj, regex, opts)
        -- ...
        return setmetatable(re_gmatch_iterator_ctx, re_gmatch_iterator_mt)
    end
end
```

Comment (spacewander) [inline lib/resty/core/regex.lua]: @thibaultcha 
Good idea! I just add this optimization.

File: lib/resty/core/regex.lua [hunk 6]
```diff
@@ -479,6 +480,76 @@ function ngx.re.find(subj, regex, opts, ctx, nth)
...
+    function ngx.re.gmatch(subj, regex, opts)
+        subj  = tostring(subj)
+
+        local compiled, compile_once, flags = re_match_compile(regex, opts)
+        if compiled == nil then
+            -- compiled_once holds the error string
+            return nil, compile_once
+        end
+
+        local re_gmatch_iterator = new_tab(0, 6)
```
Comment (thibaultcha) [inline lib/resty/core/regex.lua]: I think we can further optimize the performance here by using the same GC trick for the state as we do for the metatable:

```diff
diff --git a/lib/resty/core/regex.lua b/lib/resty/core/regex.lua
index 363dccc..fbabb33 100644
--- a/lib/resty/core/regex.lua
+++ b/lib/resty/core/regex.lua
@@ -527,6 +527,8 @@ end
 
 do
     local re_gmatch_iterator_mt = { __call = iterate_re_gmatch }
+    local re_gmatch_iterator_state = setmetatable(new_tab(0, 6),
+                                                  re_gmatch_iterator_mt)
 
     function ngx.re.gmatch(subj, regex, opts)
         subj  = tostring(subj)
@@ -537,15 +539,14 @@ do
             return nil, compile_once
         end
 
-        local re_gmatch_iterator = new_tab(0, 6)
-        re_gmatch_iterator._compiled = compiled
-        re_gmatch_iterator._compile_once = compile_once
-        re_gmatch_iterator._subj = subj
-        re_gmatch_iterator._subj_len = #subj
-        re_gmatch_iterator._flags = flags
-        re_gmatch_iterator._pos = 0
+        re_gmatch_iterator_state._compiled = compiled
+        re_gmatch_iterator_state._compile_once = compile_once
+        re_gmatch_iterator_state._subj = subj
+        re_gmatch_iterator_state._subj_len = #subj
+        re_gmatch_iterator_state._flags = flags
+        re_gmatch_iterator_state._pos = 0
 
-        return setmetatable(re_gmatch_iterator, re_gmatch_iterator_mt)
+        return re_gmatch_iterator_state
     end
 end
```

Comment (thibaultcha) [inline lib/resty/core/regex.lua]: Hm, maybe not, if we use several instances of this iterator on different subjects/regexps at once, as it seems to be a supported use-case... Too bad.

Comment (thibaultcha) [inline lib/resty/core/regex.lua]: Maybe we should have the following test to ensure the iterator is statefull and allows for the aforementioned use-case:

```
=== TEST 5: is stateful
--- http_config eval: $::HttpConfig
--- config
    location = /re {
        content_by_lua_block {
            local gmatch = ngx.re.gmatch

            local iter1 = gmatch("98", [[\d]])
            local iter2 = gmatch("12", [[\d]])

            local m1 = iter1()
            local m2 = iter2()

            ngx.say("matched iter1 (1/2): ", m1[0], " ", m1[1])
            ngx.say("matched iter2 (1/2): ", m2[0], " ", m2[1])

            m1 = iter1()
            m2 = iter2()

            ngx.say("matched iter1 (2/2): ", m1[0], " ", m1[1])
            ngx.say("matched iter2 (2/2): ", m2[0], " ", m2[1])
        }
    }
--- request
GET /re
--- response_body
matched iter1 (1/2): 9 nil
matched iter2 (1/2): 1 nil
matched iter1 (2/2): 8 nil
matched iter2 (2/2): 2 nil
--- no_error_log
[error]
```

Comment (spacewander) [inline lib/resty/core/regex.lua]: @thibaultcha 
I just added a new test case with your suggestion. Thanks!

**Comment by spacewander** [comment]
@agentzh 
Since a new release was made, do you get some spare time to review this pr again?

**Comment by agentzh** [comment]
@spacewander The test failures I'm seeing when running ngx_lua's test suite using lua-resty-core are in `t/035-gmatch.t`. For example, `TEST 19: gmatch iterator used by another request` and another test case leading to infinite loop.

File: t/re-base.t [hunk 7]
```diff
@@ -0,0 +1,226 @@
+# vim:set ft= ts=4 sw=4 et fdm=marker:
```
[the remark this answers is not in the document]
Comment (spacewander) [reply t/re-base.t]: > IMHO, errors from re_match_compile and ngx_http_lua_ffi_exec_regex do not strictly belong to ngx.re.gmatch. Maybe we could add a re-base.t later and put the test cases there?

The `re-base.t` stores common test cases related with `re_match_compile` and `ngx_http_lua_ffi_exec_regex`, added for the former suggestion from @thibaultcha.

[the remark this answers is not in the document]
Comment (agentzh) [reply t/re-base.t]: @spacewander OK, I see.

**Comment by spacewander** [comment]
@agentzh 
Thank you for your review!

There is a bug when handling empty matched string, which led `TEST 20: gmatch (empty matched string)` to infinite loop. I just submit a fix for it.

`TEST 19: gmatch iterator used by another request` fails, because with `lua-resty-core` we could share regex data across multiple requests. The implement in `lua-nginx-module` stores regex data per request, so we have to check if the iterator is used in a different request. But the one in `lua-resty-core` stores regex data in lrucache.
If we consider `be able to use iterator in other request` is a bug, we could store the reference of current request and check it in each call.

**Comment by agentzh** [comment]
@spacewander I see. It's indeed fine to update `TEST 19: gmatch iterator used by another request` for our FFI implementation after we retire the old non-FFI implementation.

I've run your `t/re-gmatch.t` test file in this branch with the Valgrind test mode and found a use-after-free memory issue when executing `TEST 14: gmatch (empty matched string)`, your newly added test block:

```
==2225== Invalid read of size 8
==2225==    at 0x4C34460: memmove (vg_replace_strmem.c:1252)
==2225==    by 0x54B76FD: lj_cconv_tv_ct (lj_cconv.c:413)
==2225==    by 0x54B623D: lj_cdata_get (lj_cdata.c:252)
==2225==    by 0x54CE089: lj_cf_ffi_meta___index (lib_ffi.c:158)
==2225==    by 0x546B1B7: lj_BC_FUNCC (in /opt/luajit21sysm/lib/libluajit-5.1.so.2.1.0)
==2225==    by 0x547EE2A: lua_resume (lj_api.c:1221)
==2225==    by 0x4E098D: ngx_http_lua_run_thread (ngx_http_lua_util.c:1010)
==2225==    by 0x4E3726: ngx_http_lua_content_by_chunk (ngx_http_lua_contentby.c:120)
==2225==    by 0x4E38C4: ngx_http_lua_content_handler_inline (ngx_http_lua_contentby.c:310)
==2225==    by 0x4E3145: ngx_http_lua_content_handler (ngx_http_lua_contentby.c:222)
==2225==    by 0x4594F8: ngx_http_core_content_phase (ngx_http_core_module.c:1173)
==2225==    by 0x453E74: ngx_http_core_run_phases (ngx_http_core_module.c:862)
==2225==    by 0x453F17: ngx_http_handler (ngx_http_core_module.c:845)
==2225==    by 0x45F5B6: ngx_http_process_request (ngx_http_request.c:1950)
==2225==    by 0x45FB7D: ngx_http_process_request_headers (ngx_http_request.c:1377)
==2225==    by 0x45FF0B: ngx_http_process_request_line (ngx_http_request.c:1049)
==2225==    by 0x460268: ngx_http_wait_request_handler (ngx_http_request.c:507)
==2225==    by 0x447240: ngx_epoll_process_events (ngx_epoll_module.c:902)
==2225==    by 0x43CA58: ngx_process_events_and_timers (ngx_event.c:252)
==2225==    by 0x44618B: ngx_single_process_cycle (ngx_process_cycle.c:332)
==2225==    by 0x41DC36: main (nginx.c:378)
==2225==  Address 0x8f73560 is 32 bytes inside a block of size 72 free'd
==2225==    at 0x4C2ED5A: free (vg_replace_malloc.c:530)
==2225==    by 0x41F3B7: ngx_destroy_pool (ngx_palloc.c:79)
==2225==    by 0x4D8697: ngx_http_lua_ffi_destroy_regex (ngx_http_lua_regex.c:2421)
==2225==    by 0x546D287: lj_vm_ffi_call (in /opt/luajit21sysm/lib/libluajit-5.1.so.2.1.0)
==2225==    by 0x54B972B: lj_ccall_func (lj_ccall.c:1161)
==2225==    by 0x54CDC1B: lj_cf_ffi_meta___call (lib_ffi.c:230)
==2225==    by 0x546B1B7: lj_BC_FUNCC (in /opt/luajit21sysm/lib/libluajit-5.1.so.2.1.0)
==2225==    by 0x547EE2A: lua_resume (lj_api.c:1221)
==2225==    by 0x4E098D: ngx_http_lua_run_thread (ngx_http_lua_util.c:1010)
==2225==    by 0x4E3726: ngx_http_lua_content_by_chunk (ngx_http_lua_contentby.c:120)
==2225==    by 0x4E38C4: ngx_http_lua_content_handler_inline (ngx_http_lua_contentby.c:310)
==2225==    by 0x4E3145: ngx_http_lua_content_handler (ngx_http_lua_contentby.c:222)
==2225==    by 0x4594F8: ngx_http_core_content_phase (ngx_http_core_module.c:1173)
==2225==    by 0x453E74: ngx_http_core_run_phases (ngx_http_core_module.c:862)
==2225==    by 0x453F17: ngx_http_handler (ngx_http_core_module.c:845)
==2225==    by 0x45F5B6: ngx_http_process_request (ngx_http_request.c:1950)
==2225==    by 0x45FB7D: ngx_http_process_request_headers (ngx_http_request.c:1377)
==2225==    by 0x45FF0B: ngx_http_process_request_line (ngx_http_request.c:1049)
==2225==    by 0x460268: ngx_http_wait_request_handler (ngx_http_request.c:507)
==2225==    by 0x447240: ngx_epoll_process_events (ngx_epoll_module.c:902)
==2225==    by 0x43CA58: ngx_process_events_and_timers (ngx_event.c:252)
==2225==    by 0x44618B: ngx_single_process_cycle (ngx_process_cycle.c:332)
==2225==    by 0x41DC36: main (nginx.c:378)
==2225==  Block was alloc'd at
==2225==    at 0x4C2DBAD: malloc (vg_replace_malloc.c:299)
==2225==    by 0x4406A5: ngx_alloc (ngx_alloc.c:22)
==2225==    by 0x41F207: ngx_malloc (ngx_palloc.c:140)
==2225==    by 0x41F446: ngx_palloc (ngx_palloc.c:123)
==2225==    by 0x4D8234: ngx_http_lua_ffi_compile_regex (ngx_http_lua_regex.c:2202)
==2225==    by 0x546D287: lj_vm_ffi_call (in /opt/luajit21sysm/lib/libluajit-5.1.so.2.1.0)
==2225==    by 0x54B972B: lj_ccall_func (lj_ccall.c:1161)
==2225==    by 0x54CDC1B: lj_cf_ffi_meta___call (lib_ffi.c:230)
==2225==    by 0x546B1B7: lj_BC_FUNCC (in /opt/luajit21sysm/lib/libluajit-5.1.so.2.1.0)
==2225==    by 0x547EE2A: lua_resume (lj_api.c:1221)
==2225==    by 0x4E098D: ngx_http_lua_run_thread (ngx_http_lua_util.c:1010)
==2225==    by 0x4E3726: ngx_http_lua_content_by_chunk (ngx_http_lua_contentby.c:120)
==2225==    by 0x4E38C4: ngx_http_lua_content_handler_inline (ngx_http_lua_contentby.c:310)
==2225==    by 0x4E3145: ngx_http_lua_content_handler (ngx_http_lua_contentby.c:222)
==2225==    by 0x4594F8: ngx_http_core_content_phase (ngx_http_core_module.c:1173)
==2225==    by 0x453E74: ngx_http_core_run_phases (ngx_http_core_module.c:862)
==2225==    by 0x453F17: ngx_http_handler (ngx_http_core_module.c:845)
==2225==    by 0x45F5B6: ngx_http_process_request (ngx_http_request.c:1950)
==2225==    by 0x45FB7D: ngx_http_process_request_headers (ngx_http_request.c:1377)
==2225==    by 0x45FF0B: ngx_http_process_request_line (ngx_http_request.c:1049)
==2225==    by 0x460268: ngx_http_wait_request_handler (ngx_http_request.c:507)
==2225==    by 0x447240: ngx_epoll_process_events (ngx_epoll_module.c:902)
==2225==    by 0x43CA58: ngx_process_events_and_timers (ngx_event.c:252)
==2225==    by 0x44618B: ngx_single_process_cycle (ngx_process_cycle.c:332)
==2225==    by 0x41DC36: main (nginx.c:378)
==2225==
```

File: t/re-gmatch.t [hunk 8]
```diff
@@ -0,0 +1,509 @@
...
+use Test::Nginx::Socket::Lua;
+use Cwd qw(cwd);
+
+#worker_connections(1014);
+#master_process_enabled(1);
+#log_level('warn');
+
+repeat_each(2);
+
+plan tests => repeat_each() * (blocks() * 4 - 5);
```
Comment (agentzh) [inline t/re-gmatch.t]: It's not a good idea to substract a larger value (`5`) from the test block average (`4`) since it will lead to a negative test plan when running a test block individually via `--- ONLY`. Please change it to an addition calculation here instead.

File: lib/resty/core/regex.lua [hunk 9]
```diff
@@ -478,6 +479,84 @@ function ngx.re.find(subj, regex, opts, ctx, nth)
     return re_match_helper(subj, regex, opts, ctx, false, nil, nth)
 end
 
+local function destroy_re_gmatch_iterator(iterator)
+    if not iterator._compile_once then
+        destroy_compiled_regex(iterator._compiled)
```
Comment (agentzh) [inline lib/resty/core/regex.lua]: Better clear the `iterator._compiled` field with `nil` here as well? Oh, and other fields too.

File: lib/resty/core/regex.lua [hunk 10]
```diff
@@ -478,6 +479,84 @@ function ngx.re.find(subj, regex, opts, ctx, nth)
...
+            return nil, compile_once
+        end
+
+        local re_gmatch_iterator = new_tab(0, 6)
+        re_gmatch_iterator._compiled = compiled
+        re_gmatch_iterator._compile_once = compile_once
+        re_gmatch_iterator._subj = subj
+        re_gmatch_iterator._subj_len = #subj
+        re_gmatch_iterator._flags = flags
+        re_gmatch_iterator._pos = 0
```
Comment (agentzh) [inline lib/resty/core/regex.lua]: The lines above look too verbose and cumbersome. I think it's nicer and also faster to use the following form instead:

```
local re_match_iterator = {
    _compiled = compiled,
    _compie_once = compile_once,
    _subj = subj,
    _subj_len = #subj,
    _flags = flags,
    _pos = 0,
}
```

File: lib/resty/core/regex.lua [hunk 11]
```diff
@@ -520,7 +520,15 @@ local function iterate_re_gmatch(self)
         rc = 1
     end
 
-    self._pos = compiled.captures[1]
+    local cp_pos = compiled.captures[1]
```
Comment (agentzh) [inline lib/resty/core/regex.lua]: We should use `tonumber()` here to reduce the number of cdata objects, which are GC objects.

File: lib/resty/core/regex.lua [hunk 12]
```diff
@@ -520,7 +520,15 @@ local function iterate_re_gmatch(self)
...
+    if cp_pos == compiled.captures[0] then
+        cp_pos = cp_pos + 1
+        if cp_pos > subj_len then
+            destroy_re_gmatch_iterator(self)
+            return collect_captures(compiled, rc, subj, flags)
```
Comment (agentzh) [inline lib/resty/core/regex.lua]: Mind you: `compiled` is already freed at this point.

**Comment by spacewander** [comment]
@agentzh 
Sorry for my carelessness.
I have just fix the mistake and run valgrind with `re-gmatch.t`.

**Comment by agentzh** [comment]
@thibaultcha Will you have another look at this please? Thanks!

**Comment by agentzh** [comment]
@spacewander Have you benchmarked the latest version on your side again? How is the result? Thanks!

File: lib/resty/core/regex.lua [hunk 13]
```diff
@@ -478,6 +479,87 @@ function ngx.re.find(subj, regex, opts, ctx, nth)
...
+    if rc == 0 then
+        if band(flags, FLAG_DFA) == 0 then
+            destroy_re_gmatch_iterator(self)
+            return nil, "capture size too small"
+        end
+
+        rc = 1
+    end
+
+    local cp_pos = tonumber(compiled.captures[1])
```
Comment (thibaultcha) [inline lib/resty/core/regex.lua]: `tonumber` is a global that hasn't been cached locally at the top of this module. Shall we add it?

File: lib/resty/core/regex.lua [hunk 14]
```diff
@@ -478,6 +479,87 @@ function ngx.re.find(subj, regex, opts, ctx, nth)
...
+            destroy_re_gmatch_iterator(self)
+            return res
+        end
+    end
+    self._pos = cp_pos
+    return collect_captures(compiled, rc, subj, flags)
+end
+
+
+do
```
Comment (thibaultcha) [inline lib/resty/core/regex.lua]: The 2 above local functions could be declared inside this `do ... end` block. What do you think?

File: lib/resty/core/regex.lua (hunk 9 above)
Comment (thibaultcha) [reply lib/resty/core/regex.lua]: Wouldn't the `iterator` table be GC'ed anyways, and so will `._compiled` be too? I am curious to hear more about this. Does it help because `._compiled` is a cdata object?

I have two suggestions here:
* if we want to unset all fields of the `iterator` table, then only 3 fields are currently unset. Should we unset them all? Or do we only want to clear cdata objects?
* we could add a test asserting that an iterator that error-ed out returns `nil` on subsequent calls (Thus effectively testing that we unset `._pos` here, as we should)

File: t/re-gmatch.t [hunk 15]
```diff
@@ -0,0 +1,509 @@
...
+false
+false
+--- error_log eval
+qr/\[TRACE\s+\d+\s+/
+--- no_error_log
+[error]
+
+
+
+=== TEST 10: unmatched name captures are false
```
Comment (thibaultcha) [inline t/re-gmatch.t]: minor typo: I think you meant `named captures`?

File: lib/resty/core/regex.lua (hunk 9 above)
Comment (agentzh) [reply lib/resty/core/regex.lua]: @thibaultcha This serves as a minor optimization by releasing the GC references sooner rather than later.

File: lib/resty/core/regex.lua (hunk 13 above)
Comment (agentzh) [reply lib/resty/core/regex.lua]: Yeah, good catch.

File: lib/resty/core/regex.lua [hunk 16]
```diff
@@ -478,6 +479,87 @@ function ngx.re.find(subj, regex, opts, ctx, nth)
...
+
+        local re_gmatch_iterator = {
+            _compiled = compiled,
+            _compile_once = compile_once,
+            _subj = subj,
+            _subj_len = #subj,
+            _flags = flags,
+            _pos = 0,
+        }
+        return setmetatable(re_gmatch_iterator, re_gmatch_iterator_mt)
```
[the remark this answers is not in the document]
Comment (agentzh) [reply lib/resty/core/regex.lua]: Agreed.

File: lib/resty/core/regex.lua (hunk 9 above)
Comment (thibaultcha) [reply lib/resty/core/regex.lua]: Gotcha. That was my hunch as well. Then I think we can avoid the `._subj = nil` and just unset `._compiled` and `._pos` maybe. Plus the mentioned test - I believe it'd be a healthy one to have.

Comment (agentzh) [reply lib/resty/core/regex.lua]: @thibaultcha Either way is fine :)

File: lib/resty/core/regex.lua [hunk 17]
```diff
@@ -478,6 +479,87 @@ function ngx.re.find(subj, regex, opts, ctx, nth)
...
+    end
+    self._pos = cp_pos
+    return collect_captures(compiled, rc, subj, flags)
+end
+
+
+do
+    local re_gmatch_iterator_mt = { __call = iterate_re_gmatch }
+
+    function ngx.re.gmatch(subj, regex, opts)
```
Comment (membphis) [inline lib/resty/core/regex.lua]:
```lua
do
    local ...

function ngx.re.gmatch(...)
    -- Lua ...

    return ...
end

end -- do
```

@spacewander How about this style? The function `ngx.re.gmatch` does not need to be indented.

File: t/re-gmatch.t [hunk 18]
```diff
@@ -0,0 +1,509 @@
...
+    }
+--- request
+GET /re
+--- response_body
+matched: hello
+matched: world
+--- error_log eval
+qr/\[TRACE   \d+ content_by_lua\(nginx\.conf:\d+\):4 loop\]/
+--- no_error_log
+[error]
```
[the remark this answers is not in the document]
Comment (spacewander) [reply t/re-gmatch.t]: Good catch!

File: lib/resty/core/regex.lua (hunk 9 above)
Comment (spacewander) [reply lib/resty/core/regex.lua]: @thibaultcha 
`._subj = nil` frees the reference to the `subj`, so it could be GCed eariler.

File: lib/resty/core/regex.lua (hunk 17 above)
Comment (spacewander) [reply lib/resty/core/regex.lua]: @membphis 
`ngx.req.get_method` in `resty.core.request.lua` is indented too. Maybe we could just follow the style.

File: lib/resty/core/regex.lua (hunk 14 above)
Comment (spacewander) [reply lib/resty/core/regex.lua]: Agree.

File: t/re-gmatch.t (hunk 15 above)
Comment (spacewander) [reply t/re-gmatch.t]: Good catch +1

**Comment by spacewander** [comment]
@thibaultcha 
Thank you for your good catch! I have fixed some style problems and add a new test to assert iterator behavior(`TEST 13: an error-ed out gmatch iterator should return nil`).

@agentzh 
The latest benchmark in my side shows that the FFI(jittable) version is still over 30% faster than the original one.
For the detail, Case 1 is 31% faster and Case 2 is 49% faster.

The travis ci fails currently because of a not-jitted failure in `shdict.t`.

**Comment by agentzh** [comment]
@spacewander That's great! Will you please add a link to `ngx.re.gmatch`'s docs to the README file of this repo under the `resty.core.regex` section below?

https://github.com/openresty/lua-resty-core/#restycoreregex

Thanks!

**Comment by spacewander** [comment]
@agentzh 
Added it.

**Comment by spacewander** [comment]
@agentzh 
Anything else is required to do?

**Comment by agentzh** [comment]
@spacewander Sorry for the delay. I'll look into this.

**Comment by agentzh** [comment]
@spacewander Merged. Thanks!
